--- result/lib/python3.11/site-packages/fanctrl.py	1969-12-31 16:00:01.000000000 -0800
+++ src/fanctrl.py	2024-05-01 11:03:44.863314224 -0700
@@ -62,7 +62,7 @@
                 stdout=subprocess.PIPE,
                 shell=True,
                 text=True,
-                executable="/bin/bash",
+                executable="bash",
             ).stdout
         self.cpu_type = "Intel" if "GenuineIntel" in cpuinfo else "AMD"
 
@@ -72,7 +72,7 @@
                 stdout=subprocess.PIPE,
                 shell=True,
                 text=True,
-                executable="/bin/bash",
+                executable="bash",
             ).stdout
         self.fan_count = faninfo.count("Fan")
         self.laptop_model = "Framework laptop 16" if self.fan_count > 1 else "Framework laptop 13"
@@ -166,7 +166,7 @@
                     stdout=subprocess.PIPE,
                     shell=True,
                     text=True,
-                    executable="/bin/bash",
+                    executable="bash",
                 ).stdout
             )
 
