--- result/lib/python3.11/site-packages/fanctrl.py	1969-12-31 16:00:01.000000000 -0800
+++ ./fanctrl.py	2024-05-01 11:17:28.320867515 -0700
@@ -138,6 +138,9 @@
         currentTemp = min(
             currentTemp, self.getMovingAverageTemperature(self.movingAverageInterval)
         )
+        if currentTemp == 0:
+            # Don't change the speed of the fan if we don't know the temp
+            return;
         minPoint = self.speedCurve[0]
         maxPoint = self.speedCurve[-1]
         for e in self.speedCurve:
@@ -222,9 +221,15 @@
     # return mean temperature over a given time interval (in seconds)
     def getMovingAverageTemperature(self, timeInterval):
         tempSum = 0
+        timeCounted = 0
         for i in range(0, timeInterval):
-            tempSum += self.temps[self._tempIndex - i]
-        return tempSum / timeInterval
+            tempVal = self.temps[self._tempIndex - i]
+            if tempVal != 0:
+                tempSum += self.temps[self._tempIndex - i]
+                timeCounted += 1
+        if timeCounted == 0:
+            return 0
+        return tempSum / timeCounted
 
     def printState(self):
         print(
